src/Type/Php/DateIntervalDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,42 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use DateInterval;
+use PhpParser\Node\Expr\StaticCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicStaticMethodReturnTypeExtension;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeUtils;
+
+class DateIntervalDynamicReturnTypeExtension implements DynamicStaticMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return DateInterval::class;
+	}
+
+	public function isStaticMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'createFromDateString';
+	}
+
+	public function getTypeFromStaticMethodCall(MethodReflection $methodReflection, StaticCall $methodCall, Scope $scope): ?Type
+	{
+		$dateTimeString = TypeUtils::getConstantStrings($scope->getType($methodCall->getArgs()[0]->value));
+
+		if (!($dateTimeString instanceof ConstantStringType)) {
+			return null;
+		}
+
+		$isValid = DateInterval::createFromDateString($dateTimeString->getValue()) !== false;
+
+		return $isValid ? new ObjectType(DateInterval::class) : new ConstantBooleanType(false);
```

Comment:
since https://github.com/phpstan/phpstan-src/pull/1979 got merged this can now be simplified to something like this (and requires a few more tests)

```suggestion
		$strings = $scope->getType($methodCall->getArgs()[0]->value)->getConstantStrings();
		
		if ($strings === []) {
		   return null;
		}
		
		if (count($strings) === 1) {
           if (DateInterval::createFromDateString($dateTimeString->getValue()) === false) {
    		 return new ConstantBooleanType(false);
           }
           return new ObjectType(DateInterval::class);
		}
		
		foreach($strings as $string) {
		  if (DateInterval::createFromDateString($dateTimeString->getValue()) === false) {
		    return null;
		  }
		}

		return new ObjectType(DateInterval::class);
```

Reply:
intresting, so if i understand this correctly, `getConstantStrings()` will return me _all possible constant strings_ with which the method could be called given the current context.
so for the example you gave earlier:

```php
if (rand(0,1)) {
  $interval = 'P1Y';
} else {
  $interval = 'P2Y';
}
```

i have two entries once `P1Y` and once `P2Y`. and were are testing both?

i think now it clicked, i was a bit confused by that.

Reply:
Correct